rfcs/0024-hard-forks.md

```diff
@@ -0,0 +1,455 @@
+
+We extend each block with an `era_id` field, which is a 32 bit integer. Each ID
+is used at most once, and corresponds to a specific set of features. They are
+not necessarily sequential, and are distinct from era numbers, which are. If we
+change what features are in a particular era, we use a new ID. Each client
+executable includes the set of global slot interval to era id mappings, and
+checks that every block it accepts has the appropriate era id for its slot. The
+blockchain SNARK certifies that the current and prior era IDs are correct. A
+block with a global slot that has no corresponding era ID is invalid.
+
+#### Feature flags
+
+In order to develop different features in parallel and avoid having painful long
+lived branches, we'll adopt feature flags for blocks and transactions. This is
+essential for disabling individual features that aren't ready yet without
+breaking other code or requiring extensive repo surgery.
+
+Feature flags will be determined statically, and the changes to types will be
+enforced by the type system, using a system inspired by Trees That Grow. This
+allows us to easily change which features are enabled, and express in the type
+system which functions do and don't interact with particular features. A sketch
+of how this will work, using GADTs:
+
+```ocaml
+
+(** Type level booleans *)
+type true_ty = Mk_true
+type false_ty = Mk_false
+
+module Static_option = struct
+  type (_, _) t =
+   | SSome : 'a -> (true_ty, 'a) t
+   | SNone : (false_ty, 'a) t
+
+  let get : ('true_ty, 'a) t -> 'a = function SSome a -> a
+
+  let map : type e. f:('a -> 'b) -> (e, 'a) t -> (e, 'b) t = fun ~f -> function
+     | SSome x -> SSome (f x)
+     | SNone -> SNone
+
+  let elim : type e. f:('a -> 'b) -> 'b -> (e, 'a) t -> 'b = fun ~f def ->
+    function
+      | SSome x -> f x
+      | SNone -> def
+end
+
+module Static_either = struct
+  type (_,_,_) t =
+    | SLeft : 'a -> (false_ty, 'a, 'b) t
+    | SRight : 'b -> (true_ty, 'a, 'b) t
+
+  let get_l : (false_ty, 'a, _) t -> 'a = function SLeft a -> a
+
+  let map_l : f:('a -> 'b) -> (false_ty, 'a, 'c) t -> (false_ty, 'b, 'c) t =
+    fun ~f -> function
+      | SLeft a -> SLeft (f a)
+
+  let get_r : (true_ty, _, 'a) t -> 'a = function SRight a -> a
+
+  let map_r : f:('b -> 'c) -> (true_ty, 'a, 'b) t -> (true_ty, 'a, 'c) t =
+    fun ~f -> function
+      | SRight a -> SRight (f a)
+
+  let bimap :
+    type w. f:('a -> 'c) -> g:('b -> 'd) -> (w, 'a, 'b) t -> (w, 'c, 'd) t =
+    fun ~f ~g -> function
+      | SLeft a -> SLeft (f a)
+      | SRight b -> SRight (g b)
+
+  let elim : f:('a -> 'c) -> g:('b -> 'c) -> (_, 'a, 'b) t -> 'c = fun ~f ~g ->
+    function
+      | SLeft a -> f a
+      | SRight a -> g a
+end
+
+(** A block structure with a transaction counter and possible change to the
+    parallel scan data structure, configured with feature flags. *)
+module Block = struct
+  type ('has_transaction_counter, 'use_new_pscan) t =
```

Comment:
nit: phantom object type here would make the true/false assignment less error prone and clearer